src/nupic/regions/SPRegion.hpp

```diff
@@ -0,0 +1,130 @@
+/* ---------------------------------------------------------------------
+ * Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2018, Numenta, Inc.  Unless you have an agreement
+ * with Numenta, Inc., for a separate license for this software code, the
+ * following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ *
+ * Author: David Keeney, Ported from Python April, 2018
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+ * Declarations for SPRegion class
+ */
+
+//----------------------------------------------------------------------
+
+#ifndef NTA_SPREGION_HPP
+#define NTA_SPREGION_HPP
+
+#include <memory> //unique_ptr
+
+#include <nupic/engine/RegionImpl.hpp>
+#include <nupic/algorithms/SpatialPooler.hpp>
+//----------------------------------------------------------------------
+
+using namespace nupic::algorithms::spatial_pooler;
+
+namespace nupic
+{
+class SPRegion  : public RegionImpl
+{
+		
+  public:
+    SPRegion(const ValueMap& params, Region *region);
+    SPRegion(BundleIO& bundle, Region* region);
+    virtual ~SPRegion();
+
+
+/* -----------  Required RegionImpl Interface methods ------- */
+
+    // Used by RegionImplFactory to create and cache
+    // a nodespec. Ownership is transferred to the caller.
+    static Spec* createSpec();
```

Comment:
does this have to be static? 
Can you make it a smart ptr?

Reply:
yes, it must be static.  It must be read before the region implementation is created.